src/crypto/public_key/elliptic/curves.js

```diff
@@ -27,107 +27,210 @@
 
 'use strict';
 
-import {ec as EC} from 'elliptic';
-import {KeyPair} from './key.js';
-import BigInteger from '../jsbn.js';
-import enums from '../../../enums.js';
-import util from '../../../util.js';
+import { ec as EC, eddsa as EdDSA } from 'elliptic';
+import { KeyPair } from './key';
+import BigInteger from '../jsbn';
+import random from '../../random';
+import config from '../../../config';
+import enums from '../../../enums';
+import util from '../../../util';
+import base64 from '../../../encoding/base64';
+
+const webCrypto = util.getWebCrypto();
+const nodeCrypto = util.getNodeCrypto();
+
+var webCurves = {}, nodeCurves = {};
+if (webCrypto && config.use_native) {
+  webCurves = {
+    'p256': 'P-256',
+    'p384': 'P-384',
+    'p521': 'P-521'
+  };
+} else if (nodeCrypto && config.use_native) {
+  var knownCurves = nodeCrypto.getCurves();
+  nodeCurves = {
+    'secp256k1': knownCurves.includes('secp256k1') ? 'secp256k1' : undefined,
+    'p256': knownCurves.includes('prime256v1') ? 'prime256v1' : undefined,
+    'p384': knownCurves.includes('secp384r1') ? 'secp384r1' : undefined,
+    'p521': knownCurves.includes('secp521r1') ? 'secp521r1' : undefined
+    // TODO add more here
+  };
+}
 
 const curves = {
   p256: {
     oid: util.bin2str([0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07]),
-    curveName: 'P-256',
-    hashName: 'SHA-256',
+    keyType: enums.publicKey.ecdsa,
```

Comment:
I don't really understand how these key types work. Why does this curve have type ecdsa if it can also be used for ecdh? It seems like ecdsa and ecdh are treated the same way in the curve generation function anyway.

Reply:
It's not to differentiate between ECDSA and ECDH, but ECDSA and EdDSA. I named it keyType because it matches the higher level keyType in the key packet (as opposed to subKeyType, which would be used to differentiate between ECDH and some other DH algorithm)

Reply:
So you basically use "ecdsa" to mean "ecdsa or ecdh" and "ecdh" to mean "only ecdh"? Instead could we just simplify this to have key types EC and Ed? It doesn't look like you ever use the ecdsa vs ecdh distinction, even though I see that curve25519 has the ecdh key type.

Reply:
Would it be acceptable to set the keyType of curve25519 to also be ecdsa? Technically speaking, it can be used for ECDSA. Just asking because this is probably the easiest fix.

Reply:
Yeah, that works